main.go

```diff
@@ -19,150 +18,173 @@ type SendPacket struct {
+const RateLoggerRate = 100
+
+// UDPServer sets up a UDP server and listens for connections
+func UDPServer(address string, hub *Hub) {
+    serverConn := SetupConnection(address)
+    defer serverConn.Close()
+
+    hub.sendChannel = make(chan SendPacket, 5)
+    addrMap := make(map[string]*Client)
+    go Sender(serverConn, hub.sendChannel)
+    buf := make([]byte, ConnectionBufferSize)
+
+    rateLoggers := make(map[string]*loggers.RateLogger)
+
+    for {
+        n, addr, err := serverConn.ReadFromUDP(buf)
+        if err != nil {
             log.WithFields(log.Fields{
                 "error": err,
             }).Fatal("UDP read error")
-		}
-		if strings.HasPrefix(string(buf[:n]), "CONNECT") {
-			split := strings.Split(string(buf[:n]), ":")
-			privateKey := split[1]
-			name := split[2]
-            log.WithFields(log.Fields{
-                "privateKey": privateKey,
-                "name": name,
-            }).Info("New client")
-			h.sendChannel <- SendPacket{addr: addr, data: []byte("HANDSHAKE")}
-			if _, ok := h.clientMap[privateKey]; !ok {
-				h.clientMap[privateKey] = make(map[string]*Client)
-			}
-			if _, ok := h.clientMap[privateKey][name]; !ok {
-				cli := &Client{addr: addr, process: make(chan []byte, 1),
-					name: name, privateKey: privateKey, h: h,
-					rateLoggers: make(map[string]*loggers.RateLogger)}
-				h.clientMap[privateKey][name] = cli
-				addrMap[string(addr.IP)+":"+string(addr.Port)] = cli
-				go cli.processor()
-			} else {
-				h.clientMap[privateKey][name].addr = addr
-				addrMap[string(addr.IP)+":"+string(addr.Port)] = h.clientMap[privateKey][name]
-			}
-		} else {
-			addrStr := string(addr.IP) + ":" + string(addr.Port)
-			if _, ok := addrMap[addrStr]; ok {
-				if _, ok := rateLoggers[addrStr]; !ok {
-					rateLoggers[addrStr] = loggers.NewRateLogger(100)
-				}
-
-				rateLoggers[addrStr].Log("Before process", log.Fields{
-					"addr": addr.String(),
-				})
-
-				select {
-				case addrMap[addrStr].process <- append([]byte{}, buf[:n]...):
-				default:
-				}
-			}
-		}
-	}
+        }
+        packet := buf[:n]
+        if strings.HasPrefix(string(packet), "CONNECT") {
+            NewClient(addr, &packet, &addrMap, hub)
+        } else {
+            ProcessDataPacket(addr, &packet, &addrMap, &rateLoggers)
+        }
+    }
+}
+
+// SetupConnection sets up the UDP server
+func SetupConnection(address string) *net.UDPConn {
+    serverAddr, err := net.ResolveUDPAddr("udp", address)
+    if err != nil {
+        panic(err)
+    }
+    serverConn, err := net.ListenUDP("udp", serverAddr)
+    if err != nil {
+        panic(err)
+    }
+    serverConn.SetReadBuffer(ConnectionBufferSize)
+    serverConn.SetWriteBuffer(ConnectionBufferSize)
+    return serverConn
+}
+
+// NewClient handles connect packets and creates client objects
+func NewClient(addr *net.UDPAddr, connectPacket *[]byte, addrMap *map[string]*Client, hub *Hub) {
+    split := strings.Split(string(*connectPacket), ":")
```

Comment:
Can the connectPacket string parsing be moved into a separate function/struct, like you did in `client.go`?